Modules/_ctypes/stgdict.c

```diff
@@ -367,40 +255,113 @@ PyCStructUnionType_update_stginfo(PyObject *type, PyObject *fields, int isStruct
+    if (!kwnames) {
+        goto error;
+    }
+    layout = PyObject_Vectorcall(
+        layout_func,
+        1 + (PyObject*[]){
+            NULL,
+            /* positional args */
+            type,
+            fields,
+            /* keyword args */
+            isStruct ? Py_True : Py_False,
+            baseinfo ? base : Py_None},
+        2 | PY_VECTORCALL_ARGUMENTS_OFFSET,
+        kwnames);
+    Py_DECREF(kwnames);
+    Py_DECREF(layout_func);
+    fields = NULL; // a borrowed reference we won't be using again
+    if (!layout) {
+        goto error;
+    }
+
+    tmp = PyObject_GetAttr(layout, &_Py_ID(align));
+    if (!tmp) {
+        goto error;
+    }
+    Py_ssize_t total_align = PyLong_AsInt(tmp);
+    Py_DECREF(tmp);
+    if (total_align < 0) {
+        if (!PyErr_Occurred()) {
+            PyErr_SetString(PyExc_ValueError,
+                            "align must be a non-negative integer");
+        }
+        goto error;
+    }
+
+    tmp = PyObject_GetAttr(layout, &_Py_ID(size));
+    if (!tmp) {
+        goto error;
+    }
+    Py_ssize_t total_size = PyLong_AsInt(tmp);
+    Py_DECREF(tmp);
+    if (total_size < 0) {
+        if (!PyErr_Occurred()) {
+            PyErr_SetString(PyExc_ValueError,
+                            "size must be a non-negative integer");
+        }
+        goto error;
+    }
+
+    format_spec_obj = PyObject_GetAttr(layout, &_Py_ID(format_spec));
+    if (!format_spec_obj) {
+        goto error;
+    }
+    Py_ssize_t format_spec_size;
+    const char *format_spec = PyUnicode_AsUTF8AndSize(format_spec_obj,
+                                                      &format_spec_size);
+    if (!format_spec) {
+        goto error;
     }
 
     if (stginfo->format) {
         PyMem_Free(stginfo->format);
         stginfo->format = NULL;
     }
+    stginfo->format = PyMem_Malloc(format_spec_size + 1);
+    if (!stginfo->format) {
+        PyErr_NoMemory();
+        goto error;
+    }
+    memcpy(stginfo->format, format_spec, format_spec_size + 1);
 
-    if (stginfo->ffi_type_pointer.elements)
-        PyMem_Free(stginfo->ffi_type_pointer.elements);
-
-    StgInfo *baseinfo;
-    if (PyStgInfo_FromType(st, (PyObject *)((PyTypeObject *)type)->tp_base,
-                           &baseinfo) < 0) {
-        return -1;
+    PyObject *layout_fields_obj = PyObject_GetAttr(layout, &_Py_ID(fields));
+    if (!layout_fields_obj) {
+        goto error;
     }
-    if (baseinfo) {
-        stginfo->flags |= (baseinfo->flags &
-                           (TYPEFLAG_HASUNION | TYPEFLAG_HASBITFIELD));
+    layout_fields = PySequence_Tuple(layout_fields_obj);
+    Py_DECREF(layout_fields_obj);
+    if (!layout_fields) {
+        goto error;
     }
-    if (!isStruct) {
-        stginfo->flags |= TYPEFLAG_HASUNION;
+    Py_CLEAR(layout);
+
+    Py_ssize_t len = PyTuple_GET_SIZE(layout_fields);
+
+    if (stginfo->ffi_type_pointer.elements) {
+        PyMem_Free(stginfo->ffi_type_pointer.elements);
```

Comment:
```suggestion
        PyMem_Free(stginfo->ffi_type_pointer.elements);
        stginfo->ffi_type_pointer.elements = NULL;
```

Just for safety (could it to some UAF issue?)

Reply:
I don't think so -- `ffi_type_pointer.elements` is set in the next block.

I guess you might get UAF with free-threading, but for that your suggestion won't work; there should probably be a lock around a bigger chunk of the code. Anyway that's not in scope for this PR.

(Also note that this piece of code is unchanged; I just added the braces.)